### ISSUE TITLE
Make IsWithinDistance an extension method

### ISSUE THREAD

**Issue by bricelam** [NetTopologySuite/GeoAPI#46]
It feels like this method doesn't belong on the interface. It could easily be implemented as an extension method.

``` csharp
public static bool IsWithinDistance(IGeometry thiz, IGeometry geom, double distance)
    => thiz.Distance(geom) <= distance;
```

**Comment by airbreather** [comment]
JTS [has it](https://github.com/locationtech/jts/blob/95fbe34bec06050e78cdf26a3d7d0f657d28fdde/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L451-L471) on `org.locationtech.jts.geom.Geometry`, and I believe that the intent of `GeoAPI.Geometries.IGeometry` was to have everything that that has.

**Comment by bricelam** [comment]
Sure, but Java doesn't support extension methods... 😄

**Comment by bricelam** [comment]
(And AFAIK JTS doesn't have separate interface definitions)

**Comment by FObermaier** [comment]
As pointed out in #30 I strongly support changing the GeoAPI interfaces to make them leaner and easier to actually implement them. Since the current set of geometry interfaces mirrors NTS classes their overall use is actually limited IMHO.

**Comment by airbreather** [comment]
One drawback to doing it exactly as proposed is that it's impossible to short-circuit when:
- [the envelopes themselves are separated by a distance that's greater than the threshold](https://github.com/NetTopologySuite/NetTopologySuite/blob/62883a95f2063cb9f614c034b6a6ccff56d74357/NetTopologySuite/Geometries/Geometry.cs#L414-L416), or
- [during the full computation, we find two points that are less than the threshold distance](https://github.com/NetTopologySuite/NetTopologySuite/blob/62883a95f2063cb9f614c034b6a6ccff56d74357/NetTopologySuite/Operation/Distance/DistanceOp.cs#L363).

The straightforward `Distance(geom) <= distance` is perfectly fine as a **default** to reduce the burden on implementers (see #67 / #68), but I definitely think that implementers should have the **option** to optimize for cases where the caller doesn't need the exact distance.

**Comment by bricelam** [comment]
One pattern we’ve used in the past is to have the extension method try to cast to an interface with a method of the same signature to allow implementers to provide an optimized version, but fallback to a slower default implementation.

**Comment by bricelam** [comment]
But that may be overkill and leaving the method on IGeography may just be the pragmatic thing to do.

**Comment by airbreather** [comment]
> have the extension method try to cast to an interface with a method of the same signature to allow implementers to provide an optimized version, but fallback to a slower default implementation.

Yeah, I mentioned that pattern on #67... it seems like we ought to just bite the bullet and switch to abstract classes per #68, though, if we can find the time to do so.
